python/plugins/processing/algs/qgis/FieldsMapper.py

```diff
@@ -37,150 +36,89 @@
-                return False
+    def initParameters(self, config=None):
+
+        class ParameterFieldsMapping(QgsProcessingParameterMatrix):
+
+            def valueAsPythonString(self, value, context):
+                return str(value)
 
         self.addParameter(ParameterFieldsMapping(self.FIELDS_MAPPING,
-                                                 self.tr('Fields mapping'),
-                                                 self.INPUT_LAYER))
-        self.addOutput(OutputVector(self.OUTPUT_LAYER,
-                                    self.tr('Refactored'),
-                                    base_input=self.INPUT_LAYER))
+                                                 description=self.tr('Fields mapping'),
+                                                 numberRows=3,
+                                                 hasFixedNumberRows=False,
+                                                 headers=['name',
+                                                          'type',
+                                                          'length',
+                                                          'precision',
+                                                          'expression']))
+        self.parameterDefinition(self.FIELDS_MAPPING).setMetadata({
+            'widget_wrapper': 'processing.algs.qgis.ui.FieldsMappingPanel.FieldsMappingWidgetWrapper'
+        })
 
     def name(self):
         return 'refactorfields'
 
     def displayName(self):
         return self.tr('Refactor fields')
 
-    def processAlgorithm(self, parameters, context, feedback):
-        layer = self.getParameterValue(self.INPUT_LAYER)
-        mapping = self.getParameterValue(self.FIELDS_MAPPING)
-        output = self.getOutputFromName(self.OUTPUT_LAYER)
-
-        layer = QgsProcessingUtils.mapLayerFromString(layer, context)
-        fields = QgsFields()
-        expressions = []
-
-        da = QgsDistanceArea()
-        da.setSourceCrs(layer.crs())
-        da.setEllipsoid(context.project().ellipsoid())
-
-        exp_context = layer.createExpressionContext()
+    def outputName(self):
+        return self.tr('Refactored')
 
+    def prepareAlgorithm(self, parameters, context, feedback):
+        mapping = self.parameterAsMatrix(parameters, self.FIELDS_MAPPING, context)
+        self.fields = QgsFields()
+        self.expressions = []
         for field_def in mapping:
-            fields.append(QgsField(field_def['name'],
-                                   field_def['type'],
-                                   field_def['length'],
-                                   field_def['precision']))
-
+            self.fields.append(QgsField(field_def['name'],
+                                        field_def['type'],
+                                        "",
+                                        field_def['length'],
+                                        field_def['precision']))
             expression = QgsExpression(field_def['expression'])
```

Comment:
Any reason in particular you dropped the setting of geometry calculator/units from the expression? This will prevent expressions like `$area` from correctly using the project ellipsoid and unit settings.